dash/development/base_component.py

```diff
@@ -167,7 +167,29 @@ def __init__(self, **kwargs):
 
             setattr(self, k, v)
 
-    def set_random_id(self):
+    def _set_random_id(self):
+        if getattr(self, "persistence", False):
+            raise RuntimeError(
+                """
+                Attempting to use an auto-generated ID with the `persistence` prop.
+                This is prohibited because persistence is tied to component IDs and
+                auto-generated IDs can easily change.
+
+                Please assign an explicit ID to this component.
+            """
+            )
+        if "dash_snapshots" in sys.modules:
+            raise RuntimeError(
+                """
+                Attempting to use an auto-generated ID in an app with `dash_snapshots`.
+                This is prohibited because snapshots saves the whole app layout,
+                including component IDs, and auto-generated IDs can easily change.
+                Callbacks referencing the new IDs will not work old snapshots.
+
+                Please assign an explicit ID to this component.
```

Comment:
Is there any way for us to print out "which" component it is? Or any other identifier (besides the ID) like the component type? If a user has a ton of components or callbacks, it can be difficult for them to find out which component the error message is referring too.

Reply:
See the error string prefix in `__init__` as an example - might be able to reuse that. the id itself might not be helpful, but at least seeing the component name and type

Reply:
good call, the callback definition will point to multiple components, including the type will help the user narrow it down.